#### Handling Imbalanced Dataset with Machine Learning

###### Fraud detection, need to predict class
###### fraud = 1

In [1]:
import pandas as pd
df=pd.read_csv('creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [2]:
df.shape

(284807, 31)

In [3]:
df['Class'].value_counts()#y#no of unique values in categorical variables

0    284315
1       492
Name: Class, dtype: int64

### The dataset is imbalanced 

In [4]:
#### Independent and Dependent Features
X=df.drop("Class",axis=1)
y=df.Class

In [5]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

###### No null data in any  column 

#### Cross Validation Like KFOLD and Hyperpaqrameter Tuning

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.model_selection import KFold
import numpy as np
from sklearn.model_selection import GridSearchCV

In [7]:
10.0 **np.arange(-2,3)#-2 to 2 except the last one

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [8]:
log_class=LogisticRegression()
grid={'C':10.0 **np.arange(-2,3),'penalty':['l1','l2']}
cv=KFold(n_splits=5,random_state=None,shuffle=False)

In [9]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.7)

###### Applying Hyper tuning of the model using gridsearchCV

In [10]:
clf=GridSearchCV(log_class,grid,cv=cv,n_jobs=-1,scoring='f1_macro')
clf.fit(X_train,y_train)

C:\Users\kishan\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [11]:
y_pred=clf.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85272    32]
 [   44    95]]
0.9991105181231933
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85304
           1       0.75      0.68      0.71       139

    accuracy                           1.00     85443
   macro avg       0.87      0.84      0.86     85443
weighted avg       1.00      1.00      1.00     85443



In [12]:
y_train.value_counts()

0    199011
1       353
Name: Class, dtype: int64

In [13]:
341/(199023+341)*100


0.17104391966453322

###### Lets Apply another classification model: Random forest

In [14]:
class_weight=dict({0:1,1:100})#defining the weight on data as it is imbalanced

In [15]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier(class_weight=class_weight)
classifier.fit(X_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

In [16]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85299     5]
 [   36   103]]
0.9995201479348805
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85304
           1       0.95      0.74      0.83       139

    accuracy                           1.00     85443
   macro avg       0.98      0.87      0.92     85443
weighted avg       1.00      1.00      1.00     85443



###### Precision,recall and F1 score have improved when compared to Logistic regresion

## Under Sampling

In [17]:
from collections import Counter
Counter(y_train)

Counter({0: 199011, 1: 353})

In [ ]:
#pip install imbalanced-learn

In [18]:
from collections import Counter
from imblearn.under_sampling import NearMiss
ns=NearMiss(0.8)
X_train_ns,y_train_ns=ns.fit_sample(X_train,y_train)
print("The number of classes before fit {}".format(Counter(y_train)))
print("The number of classes after fit {}".format(Counter(y_train_ns)))

C:\Users\kishan\Anaconda3\lib\site-packages\imblearn\utils\_validation.py:638: FutureWarning: Pass sampling_strategy=0.8 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


The number of classes before fit Counter({0: 199011, 1: 353})
The number of classes after fit Counter({0: 441, 1: 353})


###### So the number of dataset rows reduced from 199011 to 441 for class 0

###### now lets apply random forest on downsampled dataset

In [19]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [20]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[60670 24634]
 [   13   126]]
0.7115386866097866
              precision    recall  f1-score   support

           0       1.00      0.71      0.83     85304
           1       0.01      0.91      0.01       139

    accuracy                           0.71     85443
   macro avg       0.50      0.81      0.42     85443
weighted avg       1.00      0.71      0.83     85443



###### Precision is very less for class 1 and accuracy also goes down, As a result avoid undersampling

## Over Sampling

In [21]:
from imblearn.over_sampling import RandomOverSampler

In [22]:
os=RandomOverSampler(0.75)
X_train_ns,y_train_ns=os.fit_sample(X_train,y_train)#applying oversampling on minor class
print("The number of classes before fit {}".format(Counter(y_train)))
print("The number of classes after fit {}".format(Counter(y_train_ns)))

C:\Users\kishan\Anaconda3\lib\site-packages\imblearn\utils\_validation.py:638: FutureWarning: Pass sampling_strategy=0.75 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


The number of classes before fit Counter({0: 199011, 1: 353})
The number of classes after fit Counter({0: 199011, 1: 149258})


###### 149258/199011~= 0.75

In [23]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [26]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85299     5]
 [   34   105]]
0.9995435553526912
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85304
           1       0.95      0.76      0.84       139

    accuracy                           1.00     85443
   macro avg       0.98      0.88      0.92     85443
weighted avg       1.00      1.00      1.00     85443



#### SMOTETomek

In [28]:
from imblearn.combine import SMOTETomek

In [29]:
os=SMOTETomek(0.75)
X_train_ns,y_train_ns=os.fit_sample(X_train,y_train)
print("The number of classes before fit {}".format(Counter(y_train)))
print("The number of classes after fit {}".format(Counter(y_train_ns)))

C:\Users\kishan\Anaconda3\lib\site-packages\imblearn\utils\_validation.py:638: FutureWarning: Pass sampling_strategy=0.75 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


The number of classes before fit Counter({0: 199011, 1: 353})
The number of classes after fit Counter({0: 198326, 1: 148573})


In [32]:
148573/198326

0.7491352621441465

In [30]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [33]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85293    11]
 [   28   111]]
0.9995435553526912
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85304
           1       0.91      0.80      0.85       139

    accuracy                           1.00     85443
   macro avg       0.95      0.90      0.93     85443
weighted avg       1.00      1.00      1.00     85443



#### Ensemble Techniques

In [34]:
from imblearn.ensemble import EasyEnsembleClassifier

In [36]:
easy=EasyEnsembleClassifier()
easy.fit(X_train,y_train)

EasyEnsembleClassifier()

In [37]:
y_pred=easy.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[82637  2667]
 [   12   127]]
0.9686457638425617
              precision    recall  f1-score   support

           0       1.00      0.97      0.98     85304
           1       0.05      0.91      0.09       139

    accuracy                           0.97     85443
   macro avg       0.52      0.94      0.54     85443
weighted avg       1.00      0.97      0.98     85443



##### precision for class 1 has reduced, which is not that good 